File: SingleTop/py/singletop.py
#!/usr/bin/env python3
# Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

"""What was produced together with the top: the recoil quark, the associated W, or the b?

    python3 singletop.py ../../Common/fixtures/singletop.json
    python3 singletop.py step3.root -n 2

The C++ twin is cpp/SingleTop.cc and prints the same lines.
"""

import argparse
import sys

from TruthGraphAnalysis.Common.exampleSupport import eventNumber, graphsFrom, isParton, pt


def run(graph, out=sys.stdout):
    """Prints one line per object of interest."""
    for t in graph.signalParticles():
        for partner in graph.productionSiblings(t):
            a = abs(graph.pdgId(partner))
            kind = ("associated W" if a == 24 else "top" if a == 6 else "b" if a == 5
                    else "recoil quark" if isParton(a) else "other")
            out.write("singletop: top with %s %d pt %.2f GeV\n" % (kind, graph.pdgId(partner), pt(graph.p4(partner))))


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("-n", "--maxEvents", type=int, default=-1)
    parser.add_argument("inputs", nargs="+", help="an EDM file, or one or more JSON dumps")
    args = parser.parse_args()
    for graph in graphsFrom(args.inputs, args.maxEvents):
        print("== event %s" % eventNumber(graph))
        run(graph)


if __name__ == "__main__":
    main()

File: SingleTop/cpp/SingleTop.cc
// Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

#include "TruthGraphAnalysis/SingleTop/cpp/SingleTop.h"

#include <cmath>
#include <string>
#include <vector>

#include "PhysicsTools/TruthInfo/interface/Branch.h"
#include "PhysicsTools/TruthInfo/interface/TruthLevels.h"
#include "TruthGraphAnalysis/Common/interface/ExampleSupport.h"

namespace tga::singletop {

  using tga::fixed;

  void run(truth::Graph const& graph, std::ostream& out) {
    for (auto const& t : graph.signalParticles()) {
      for (auto const& partner : t.productionSiblings()) {
        const int32_t a = std::abs(partner.pdgId());
        // A top partner is ttbar, not single top; the top is a parton by its pdgId, so it is
        // named before the parton test.
        const char* kind = a == 24              ? "associated W"
                           : a == 6             ? "top"
                           : a == 5             ? "b"
                           : truth::isParton(a) ? "recoil quark"
                                                : "other";
        out << "singletop: top with " << kind << " " << partner.pdgId() << " pt " << fixed(partner.momentum().pt())
            << " GeV\n";
      }
    }
  }

}  // namespace tga::singletop
